// leftovers-again/lib/model/pokebarn.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _pokemon = require('./pokemon');

var _pokemon2 = _interopRequireDefault(_pokemon);

var _log = require('../log');

var _log2 = _interopRequireDefault(_log);

var _pokeutil = require('../pokeutil');

var _pokeutil2 = _interopRequireDefault(_pokeutil);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Pokebarn = function () {
  function Pokebarn() {
    _classCallCheck(this, Pokebarn);

    this.allmon = [];
  }

  _createClass(Pokebarn, [{
    key: 'all',
    value: function all() {
      return this.allmon;
    }

    /**
     * Create a new Pokemon and save it to this.allmon
     * 
     * @param  {String} ident  The ident
     * @param  {String} details  The details
     * 
     * @return {Object<Pokemon>}  The mon you've created
     */

  }, {
    key: 'create',
    value: function create(ident, details) {
      var res = new _pokemon2.default(ident, details);
      this.allmon.push(res);
      return res;
    }

    /**
     * Get a specific mon from your barn.
     * 
     * @param  {String} ident  The ident
     * 
     * @return {Object<Pokemon>}  The mon you're looking for
     */

  }, {
    key: 'find',
    value: function find(ident) {
      var searchFor = _pokeutil2.default.identWithoutPosition(ident);

      var matches = this.allmon.filter(function (mon) {
        return mon.ident === searchFor;
      });
      if (matches.length > 1) {
        _log2.default.error('Found multiple mons with the same ident! o fuck');
        _log2.default.error(matches);
      }
      return matches[0];
    }

    /**
     * Get a specific mon from your barn, or create it.
     * 
     * @param  {String} ident  The ident
     * @param  {String} details  The details
     * 
     * @return {Object<Pokemon>}  The mon you're looking for
     */

  }, {
    key: 'findOrCreate',
    value: function findOrCreate(ident, details) {
      var mon = this.find(ident);
      if (mon) return mon;
      return this.create(ident, details);
    }

    /**
     * Sometimes Pokemon get replaced. Like when Zoroark comes to town.
     *
     * @param  {[type]} ident     The ident, ex. 'p2a: Sabbs'
     * @param  {[type]} details   The details, ex. 'Sableye-Mega, M'
     * @param  {[type]} condition [description]
     * @return {[type]}           [description]
     */

  }, {
    key: 'replace',
    value: function replace(ident, details, condition) {
      // Log.debug(`replace call: ${ident}|${details}|${condition}`);
      var pos = _pokeutil2.default.identToPos(ident);
      var replaced = this.findByPos(pos);
      var idx = this.allmon.indexOf(replaced);
      if (idx >= 0) {
        // just remove this guy for now. it's kinda too complicated to
        // try to remember the old info that we're losing (ex. if it
        // was Zoroark, you know the id of an unseen Pokemon)
        this.allmon.splice(idx, 1);
      } else {
        _log2.default.error('Couldnt find the thing we want to replace.');
      }

      var updated = this.findOrCreate(ident, details);

      if (condition) {
        updated.useCondition(condition);
      } else if (replaced.condition) {
        // condition was null, so we probs got details / forme change
        // in that case, use the condition we had before? hppct/conditions
        // probably did not change, but maxhp probably did.
        updated.useCondition(replaced.condition);
      }

      if (details) {
        updated.useDetails(details, true);
      }

      return updated;
    }

    /**
     * Find a Pokemon by its position, ex. 'p2a'
     * @param  {String} pos The position of the Pokemon.
     * @return {Pokemon} The Pokemon object.
     */

  }, {
    key: 'findByPos',
    value: function findByPos(pos) {
      return this.allmon.find(function (mon) {
        return mon.position === pos;
      });
    }
  }]);

  return Pokebarn;
}();

exports.default = Pokebarn;
