// leftovers-again/lib/model/timer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/**
 * @TODO documentation
 */
var Timer = function () {
  function Timer() {
    _classCallCheck(this, Timer);

    this.timeout = null;
  }

  _createClass(Timer, [{
    key: "after",
    value: function after(cb, seconds) {
      this.cb = cb;
      this.seconds = seconds;
      this.ping(); // cancel existing timeout
      this.timeout = setTimeout(cb, seconds);
    }
  }, {
    key: "ping",
    value: function ping() {
      if (this.timeout) {
        clearTimeout(this.cb, this.seconds);
      }
    }
  }]);

  return Timer;
}();

exports.default = Timer;
